### ISSUE TITLE
[Flaky Test] integration-master: admissionwebhook.TestWebhookTimeoutWithoutWatchCache

### ISSUE THREAD

**Issue by mkorbi** [kubernetes/kubernetes#98606]
**Which jobs are flaking**:

- [integration-master](https://testgrid.k8s.io/sig-release-master-blocking#integration-master)

**Which test(s) are flaking**:
- k8s.io/kubernetes/test/integration/apiserver/admissionwebhook.TestWebhookTimeoutWithoutWatchCache
- k8s.io/kubernetes/test/integration/apiserver/admissionwebhook.TestWebhookTimeoutWithoutWatchCache/timed_out_client_requests_skip_later_mutating_webhooks_(regardless_of_failure_policy)_and_fail

**Testgrid link**:
https://testgrid.k8s.io/sig-release-master-blocking#integration-master

**Reason for failure**:
```
W0130 14:27:45.467026  118540 warnings.go:70] flowcontrol.apiserver.k8s.io/v1beta1 FlowSchema is deprecated in v1.23+, unavailable in v1.26+
W0130 14:27:45.470851  118540 warnings.go:70] flowcontrol.apiserver.k8s.io/v1beta1 FlowSchema is deprecated in v1.23+, unavailable in v1.26+
I0130 14:27:46.201015  118540 controller.go:132] OpenAPI AggregationController: action for item : Nothing (removed from the queue).
I0130 14:27:46.201050  118540 controller.go:132] OpenAPI AggregationController: action for item k8s_internal_local_delegation_chain_0000000000: Nothing (removed from the queue).
I0130 14:27:46.209009  118540 storage_scheduling.go:132] created PriorityClass system-node-critical with value 2000001000
I0130 14:27:46.233311  118540 storage_scheduling.go:132] created PriorityClass system-cluster-critical with value 2000000000
I0130 14:27:46.233341  118540 storage_scheduling.go:148] all system priority classes are created successfully or already exist.
W0130 14:27:46.291937  118540 lease.go:233] Resetting endpoints for master service "kubernetes" to [127.0.0.1]
E0130 14:27:46.293134  118540 controller.go:223] unable to sync kubernetes service: Endpoints "kubernetes" is invalid: subsets[0].addresses[0].ip: Invalid value: "127.0.0.1": may not be in the loopback range (127.0.0.0/8)
```
and
```
 === SKIP: test/integration/scheduler_perf TestSchedule100Node3KPods (0.00s)
    scheduler_test.go:74: Skipping because we want to run short tests
=== Failed
=== FAIL: test/integration/apiserver/admissionwebhook TestWebhookTimeoutWithoutWatchCache/timed_out_client_requests_skip_later_mutating_webhooks_(regardless_of_failure_policy)_and_fail (3.03s)
W0130 14:27:50.495109  118540 warnings.go:70] admissionregistration.k8s.io/v1beta1 MutatingWebhookConfiguration is deprecated in v1.16+, unavailable in v1.22+; use admissionregistration.k8s.io/v1 MutatingWebhookConfiguration
W0130 14:27:50.497392  118540 warnings.go:70] admissionregistration.k8s.io/v1beta1 ValidatingWebhookConfiguration is deprecated in v1.16+, unavailable in v1.22+; use admissionregistration.k8s.io/v1 ValidatingWebhookConfiguration
I0130 14:27:53.503357  118540 trace.go:205] Trace[804584285]: "Call mutating webhook" configuration:admission.integration.test-2,webhook:admission.integration.test.0.mutating-1-5s,resource:/v1, Resource=pods,subresource:,operation:CREATE,UID:cbc1fe6d-3939-4587-8349-091a2d830c19 (30-Jan-2021 14:27:50.506) (total time: 2996ms):
Trace[804584285]: [2.99655491s] [2.99655491s] END
W0130 14:27:53.503449  118540 dispatcher.go:170] Failed calling webhook, failing open admission.integration.test.0.mutating-1-5s: failed calling webhook "admission.integration.test.0.mutating-1-5s": Post "https://127.0.0.1:40515/mutating/1/5s?timeout=3s": context deadline exceeded
I0130 14:27:53.512910  118540 tlsconfig.go:255] Shutting down DynamicServingCertificateController
I0130 14:27:53.512934  118540 dynamic_serving_content.go:145] Shutting down serving-cert::/tmp/kubernetes-kube-apiserver186968915/apiserver.crt::/tmp/kubernetes-kube-apiserver186968915/apiserver.key
I0130 14:27:53.512954  118540 dynamic_cafile_content.go:182] Shutting down client-ca-bundle::/tmp/kubernetes-kube-apiserver186968915/client-ca.crt
I0130 14:27:53.513004  118540 secure_serving.go:241] Stopped listening on 127.0.0.1:38969
E0130 14:27:53.516347  118540 controller.go:184] Get "https://127.0.0.1:38969/api/v1/namespaces/default/endpoints/kubernetes": dial tcp 127.0.0.1:38969: connect: connection refused
    --- FAIL: TestWebhookTimeoutWithoutWatchCache/timed_out_client_requests_skip_later_mutating_webhooks_(regardless_of_failure_policy)_and_fail (3.03s)
=== FAIL: test/integration/apiserver/admissionwebhook TestWebhookTimeoutWithoutWatchCache (17.63s) 
```
**Anything else we need to know**:
- Spyglass: https://prow.k8s.io/view/gcs/kubernetes-jenkins/logs/ci-kubernetes-integration-master/1355517716575817729

**Comment by rkilingr** [comment]
Can confirm this https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/98637/pull-kubernetes-integration/1357033477781852160

**Comment by sxllwx** [comment]
The kube-apiserver processing flow is as follows, (non-critical paths are omitted)：

1. ... // other filters
2. [request_deadline_filter](https://github.com/kubernetes/kubernetes/blob/97d40890d00acf721ecabb8c9a6fec3b3234b74b/staging/src/k8s.io/apiserver/pkg/endpoints/filters/request_deadline.go#L55) parse the http client timeout setting, sets the Ctx with Deadline in the http.Request
3. ...// other filters
4. [timeout_filter](https://github.com/kubernetes/kubernetes/blob/97d40890d00acf721ecabb8c9a6fec3b3234b74b/staging/src/k8s.io/apiserver/pkg/server/filters/timeout.go#L116) select the http.Request.Context().Done()

Pay attention to the processing logic in `timeout-filter`. Although the Context of http.Request has been cancelled, the Mutex on baseTimeoutWriter needs to be obtained between entering the [timeout](https://github.com/kubernetes/kubernetes/blob/97d40890d00acf721ecabb8c9a6fec3b3234b74b/staging/src/k8s.io/apiserver/pkg/server/filters/timeout.go#L226) function. The inner-handler may have written `http.Header`, ready to write `response.Body`, but the current code flow thinks that it should be Panic quickly, and it is recovered by [http2.runHandler](https://github.com/golang/net/blob/5f4716e94777e714bc2fb3e3a44599cb40817aac/http2/server.go#L2137) and sends http2.Reset to the Client. As a result, a `stream error: stream ID xxx; INTERNAL_ERROR` appears on the client side.

> If the time gap between the innerHandler processing time and the Client timeout setting is too small, this strange phenomenon will occur.

**Comment by BenTheElder** [comment]
Currently seeing it in periodic and in presubmit for `TestWebhookTimeoutWithWatchCache`

**Comment by sxllwx** [comment]
> Currently seeing it in periodic and in presubmit for `TestWebhookTimeoutWithWatchCache`

Regardless of WithCache or WithoutCache, [the value of this Timeout setting](https://github.com/kubernetes/kubernetes/blob/9e5fcc49ec568f222a5274f443903d89fec3e591/test/integration/apiserver/admissionwebhook/timeout_test.go#L127) here is not reasonable.

**Comment by palnabarun** [comment]
PS: There were some recent failures: https://prow.k8s.io/view/gs/kubernetes-jenkins/logs/ci-kubernetes-integration-master/1367490864527446016

**Comment by sxllwx** [comment]
> #98797 may be one fix, #99259 may also be necessary (thanks for opening these @sxllwx)

I believe that #98797 can effectively repair the issue, but currently #99259  is not the key PR to fix the issue.
